src/tactic/norm_cast.lean

```diff
@@ -24,269 +43,482 @@ try_for 1000 (mk_instance e)
 do
-    (a, b) ← is_eq ty,
-    α ← infer_type a,
-    new_ty ← to_expr ``(%%b = %%a),
-    f ← to_expr ``(@eq.symm %%α %%a %%b),
-    return (new_ty, ⇑f)
+  (a, b) ← is_eq ty,
+  let new_ty := ``(%%b = %%a),
+  let new_e := ``(eq.symm %%e),
+  return (new_ty, new_e)
 
-meta def flip_iff (ty : expr) : tactic (expr × (expr → expr)) :=
+meta def flip_iff (ty e : expr) : option (pexpr × pexpr) :=
+do
+  (a, b) ← is_iff ty,
+  let new_ty := ``(%%b ↔ %%a),
+  let new_e := ``(iff.symm %%e),
+  return (new_ty, new_e)
+
+meta def flip : expr → expr → option (pexpr × pexpr)
+| (pi n bi d b) e := do
+  (b', e') ← flip b (expr.lift_vars e 0 1 (var 0)),
+  let d' := pexpr.of_expr d,
+  let new_ty := pi n bi d' b',
+  let new_e := lam n bi d' e',
+  (new_ty, new_e)
+| e ty := flip_eq e ty <|> flip_iff e ty
+
+meta def reverse (ty e : expr) : tactic (expr × expr) :=
 do
-    (a, b) ← is_iff ty,
-    new_ty ← to_expr ``(%%b ↔ %%a),
-    f ← to_expr ``(@iff.symm %%a %%b),
-    return (new_ty, ⇑f)
+  (new_ty, new_e) ← flip ty e,
+  new_ty ← to_expr new_ty,
+  new_e ← to_expr new_e,
+  return (new_ty, new_e)
+
+meta def contains_coe' (e : expr) : bool :=
+e.contains_constant name.is_coe
+
+meta def is_coe : expr → bool
+| (expr.const n _) := n.is_coe
+| _ := ff
 
 end expr
 
 namespace norm_cast
+
 open tactic expr
 
-private meta def new_name (n : name) : name := name.mk_string "reversed" n
-
-private meta def aux_after_set (tac : expr → tactic (expr × (expr → expr))) :
-    expr → tactic (expr × (expr → expr))
-| (pi n bi d b) := do
-    uniq_n ← mk_fresh_name,
-    let b' := b.instantiate_var (local_const uniq_n n bi d),
-    (b', f) ← aux_after_set b',
-    return $ (
-        pi n bi d $ b'.abstract_local uniq_n,
-        λ e, lam n bi d $ ( f $ e (local_const uniq_n n bi d) ).abstract_local uniq_n
-    )
-| ty := tac ty
-
-mk_simp_attribute push_cast "The `push_cast` simp attribute uses `move_cast` lemmas in the \"forward\" direction,
-to move casts toward the leaf nodes of the expression."
+--meta def new_name (n : name) : name := name.mk_string "reversed" n
 
-/-- Called after the `move_cast` attribute is applied to a declaration. -/
-private meta def after_set (decl : name) (prio : ℕ) (pers : bool) : tactic unit :=
-do
-    (declaration.thm n l ty e) ← get_decl decl | failed,
-    let tac := λ ty, (flip_eq ty <|> flip_iff ty),
-    (ty', f) ← aux_after_set tac ty,
-    let e' := task.map f e,
-    let n' := new_name n,
-    add_decl (declaration.thm n' l ty' e'),
-    simp_attr.push_cast.set decl () tt
+inductive lemma_type
+| Elim : lemma_type
+| Move : lemma_type
+| Squash : lemma_type
 
-private meta def mk_cache : list name → tactic simp_lemmas :=
-monad.foldl simp_lemmas.add_simp simp_lemmas.mk
+open lemma_type
 
-/--
-This is an attribute for simplification rules that are
-used to normalize casts.
+def lemma_type_to_string : lemma_type → string
+| Elim := "elim_cast"
+| Move := "move_cast"
+| Squsah := "squash_cast"
```

Comment:
```suggestion
| Squash := "squash_cast"
```

Reply:
This fried my brain until I realized it's compiling only because "Squsah" is the last pattern and is therefore seen as only a variable name

Reply:
If I hadn't made this same mistake before, I would feel the same!